app/Libraries/Beatmapset/NominateBeatmapset.php

```diff
@@ -0,0 +1,213 @@
+        ];
+
+        return $config;
+    }
+
+    private static function nominationCount(array $nominationsByType, string $type, string $ruleset): int
+    {
+        return count(array_filter($nominationsByType[$type], fn ($item) => $item === $ruleset));
+    }
+
+    public function handle()
+    {
+        $this->assertValidState();
+        $this->assertRulesetNomination();
+
+        $nomination = $this->beatmapset->beatmapsetNominations()->current()->where('user_id', $this->user->getKey());
+        if (!$nomination->exists()) {
+            $this->beatmapset->getConnection()->transaction(function () {
+                $eventParams = [
+                    'comment' => ['modes' => $this->nominatedRulesets->toArray()],
+                    'type' => BeatmapsetEvent::NOMINATE,
+                    'user_id' => $this->user->getKey(),
+                ];
+
+                $event = $this->beatmapset->events()->create($eventParams);
+                $this->beatmapset->beatmapsetNominations()->create([
+                    'event_id' => $event->getKey(),
+                    'modes' => $eventParams['comment']['modes'],
+                    'user_id' => $this->user->getKey(),
+                ]);
+
+                if ($this->shouldQualify()) {
+                    $beatmapset = $this->beatmapset->lockForUpdate()->find($this->beatmapset->getKey());
+                    // Sanity check: something changed main ruleset after the qualify check.
+                    if (
+                        $beatmapset->eligible_main_rulesets === null
+                            || count($beatmapset->eligible_main_rulesets) !== 1
+                            || $beatmapset->eligible_main_rulesets[0] !== $this->beatmapset->mainRuleset()
+                    ) {
+                        throw new InvariantException('main ruleset has changed.');
+                    }
+
+                    $beatmapset->qualify($this->user);
+                } else {
+                    (new BeatmapsetNominate($this->beatmapset, $this->user))->dispatch();
+                }
+            });
+        }
+
+        $this->beatmapset->refresh();
+        $this->beatmapset->refreshCache();
+    }
+
+    private function assertValidState(): void
+    {
+        if (!$this->beatmapset->isPending()) {
+            throw new InvariantException(osu_trans('beatmaps.nominations.incorrect_state'));
+        }
+
+        if ($this->beatmapset->hype < $this->beatmapset->requiredHype()) {
+            throw new InvariantException(osu_trans('beatmaps.nominations.not_enough_hype'));
+        }
+
+        // check if there are any outstanding issues still
+        if ($this->beatmapset->beatmapDiscussions()->openIssues()->count() > 0) {
+            throw new InvariantException(osu_trans('beatmaps.nominations.unresolved_issues'));
+        }
+    }
+
+    private function assertRulesetNomination()
+    {
+        // NOTE: This assumption is only valid for the current nomination rules.
```

Comment:
similar note should probably also in the `requiredNominationsConfig`?

Because that's where it looks like the number is configurable but it's actually hardcoded/assumed in quite a few places (at least here and nominator's `userCanNominateMode` from a quick glance).

Reply:
the assumption should be gone now 🤔

Reply:
and the assumption is back 😬